### ISSUE TITLE
Update Angular.js

### ISSUE THREAD

**Issue by csillag** [hypothesis/h#2098]
Currently, we are using an obsolete release of Angular.js: v1.2.25. It's not even top of the 1.2 branch, because that would be v1.2.28.

Meanwhile, the current stable release is v1.3.15, and 1.4 is just around the corner, too. (Later version is 1.4.0-beta.6, which is probably a release candidate.)

IIRC, we already tried to upgrade earlier (to get rid of some of the errors), but ran into some unexpected problems (because of our complex and slightly non-standard usage of Angular), so the effort was abandoned.

We should pick it up and finish it. If we do this soon, I suggest going with v1.3.15; if we wait a little bit more, then we could go directly to 1.4. (But then we should go up at least to 1.2.28 in the time being.)

**Comment by tilgovi** [comment]
> because of our complex and slightly non-standard usage of Angular)

This is the second time you've alluded to this recently. Can you please show me where or how we are doing something "non-standard"?

**Comment by csillag** [comment]
> Can you please show me where or how we are doing something "non-standard"?

The [docs](https://code.angularjs.org/1.2.28/docs/api/ng/type/$rootScope.Scope#$digest) say that

> Usually, you don't call `$digest()` directly in controllers or in directives. Instead, you should call `$apply()` (typically from within a directives), which will force a `$digest()`.

However, we are calling `$digest()` in a million places of our code.

**Comment by tilgovi** [comment]
I think your claim is total FUD.

We use it in five places, according to my grepping.

In deep-count it's used deliberately to restrict the affected scopes for performance sensitive code.
In annotation it's used for the timestamp, similarly, because it's completely local information.
In thread it's used for asynchronous rendering, again in a totally local way for performance.

That only leaves two places that I think should probably be changed: app-controller's streamer updates and annotation-ui-sync. What we have is effectively an `$apply` in this places for the purposes of dirty checking because these are at the app and root scope, respectively, so it's a (near) complete digest of all scopes. For exception handling and logging, we should be using $apply in these cases.

Anyway, guidelines are there as guidelines. Where we use `$digest` it's mostly intentional and for good reason.

**Comment by csillag** [comment]
Technically, I don't think it's accurate to characterize my observation as FUD, because that term generally implies a level of wilfulness and a strategic attempt. In my case, it's an inaccurate perception, at worst.

Here are a few experiences my perception was based on:
 * I remember a time when @gergely-ujvari have spent 3+ days (maybe a week) with trying to figure out how our various `$apply` / `$digest` calls work together. The only explanation I can imagine for this is that it was somehow less then perfectly clean and well organized. Maybe it's better now; then I guess it's partially the result of this past effort.
 * I also remember a time when we tried to update Angular, but our attempt was frustrated by some collisions of `$apply` / `$digest` calls. IIRC, our conclusion at that time was that the internal changes between angular versions broke our product because we are somehow meddling with internal things in ways we are not supposed to.
 * Just a few weeks back, we also hit a unexplained `$digest` problem in the markdown directive. 

This is why I got the impression that our `$digest` usage could be improved.

If it's no longer that bad, or it never was, then I am happy to be proven wrong.
